File: hphp/hack/src/hhbc/hhas_param.ml
(*
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

type t = {
  param_name: string;
  param_is_variadic: bool;
  param_is_inout: bool;
  param_user_attributes: Hhas_attribute.t list;
  param_type_info: Hhas_type_info.t option;
  param_default_value: (Label.t * Tast.expr) option;
}

let make
    param_name
    param_is_variadic
    param_is_inout
    param_user_attributes
    param_type_info
    param_default_value =
  {
    param_name;
    param_is_variadic;
    param_is_inout;
    param_user_attributes;
    param_type_info;
    param_default_value;
  }

let name p = p.param_name

let is_variadic p = p.param_is_variadic

let is_inout p = p.param_is_inout

let user_attributes p = p.param_user_attributes

let type_info p = p.param_type_info

let default_value p = p.param_default_value

let without_type p =
  match p.param_type_info with
  | Some ti ->
    let ut = Hhas_type_info.user_type ti in
    let tc = Hhas_type_constraint.make None [] in
    let ti = Hhas_type_info.make ut tc in
    { p with param_type_info = Some ti }
  | None -> p

let with_name name p = { p with param_name = name }

File: hphp/hack/src/hhbc/hhas_param.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.
#![feature(rustc_private)]
use hhas_attribute_rust::HhasAttribute;
use hhas_type::{constraint, Info};
use label_rust::Label;
use oxidized::ast as tast;
extern crate bitflags;

#[derive(Clone, Debug)]
pub struct HhasParam {
    pub name: String,
    pub is_variadic: bool,
    pub is_inout: bool,
    pub user_attributes: Vec<HhasAttribute>,
    pub type_info: Option<Info>,
    pub default_value: Option<(Label, tast::Expr)>,
}

impl HhasParam {
    pub fn replace_default_value_label(&mut self, new_label: Label) {
        if let Some((label, _)) = self.default_value.as_mut() {
            *label = new_label;
        }
    }

    pub fn without_type(&mut self) {
        self.type_info.as_mut().map(|ti| {
            ti.type_constraint = constraint::Type::default();
        });
    }

    pub fn with_name(&mut self, name: String) {
        self.name = name;
    }
}
